--- SubProcess.pm.old	Wed Dec 22 19:02:20 2004
+++ SubProcess.pm	Thu Feb 17 08:54:57 2005
@@ -15,6 +15,12 @@
 # We run the actual DB connection here
 use DBI;
 
+# Useful DBI constants
+use DBI qw(:sql_types);
+
+# Just for debugging
+#use Data::Dumper;
+
 # Autoflush to avoid weirdness
 select((select(STDERR), $| = 1)[0]);
 select((select(STDOUT), $| = 1)[0]);
@@ -386,7 +392,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -493,7 +500,8 @@
 			} else {
 				# Execute the query
 				try {
-					$rows_affected = $sth->execute( @{ $data->{placeholders} } );
+					$self->bind_placeholders($sth, $data->{placeholders});
+					$rows_affected = $sth->execute();
 				} catch Error with {
 					die $sth->errstr;
 				};
@@ -605,7 +613,8 @@
 		} else {
 			# Execute the query
 			try {
-				$rows_affected = $sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$rows_affected = $sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -664,7 +673,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -767,7 +777,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -894,7 +905,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -1012,7 +1024,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -1114,7 +1127,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -1195,7 +1209,8 @@
 		} else {
 			# Execute the query
 			try {
-				$sth->execute( @{ $data->{placeholders} } );
+				$self->bind_placeholders($sth, $data->{placeholders});
+				$sth->execute();
 			} catch Error with {
 				die $sth->errstr;
 			};
@@ -1275,6 +1290,55 @@
 
 	# Print it!
 	print STDOUT @$outdata;
+}
+
+sub bind_placeholders {
+	#
+	# Caveats:
+	#
+	#  -- $sth->bind_param() errors
+	#     when you have more
+	#     placeholders defined
+	#     than used
+	#     ...But doesn't execute()
+	#     have same behaviour?!
+	#     
+	#     a solution: maybe counting
+	#     "?"'s in query
+	#
+	#
+	#  -- Original error catching
+	#     is maybe (well, i some case, sure - look up) broken
+	#
+	
+	#open(FH,">db.txt");
+	#print FH Dumper(@_);
+	#return;
+	
+	
+	my $self = shift;
+	
+	my $sth = shift;
+	
+	my $placeholders = shift;
+	
+	
+	my $i = 0;
+	
+	foreach (@$placeholders) {
+		$i++;
+		
+		#print FH "$i -- $_\n";
+		
+		if (defined && /^\d+$/) {
+			$sth->bind_param($i, $_, SQL_INTEGER);
+		}
+		else {
+			$sth->bind_param($i, $_);
+		}
+	}
+	
+	#close(FH);
 }
 
 1;
